pennylane_lightning/core/lightning_interpreter.py

```diff
@@ -0,0 +1,182 @@
+        """Interpret a given measurement equation."""
+        if "mcm" == eqn.primitive.name[-3:]:
+            raise NotImplementedError(
+                "LightningInterpreter does not yet support postprocessing mcms"
+            )
+        return super().interpret_measurement_eqn(eqn)
+
+    def interpret_measurement(self, measurement):
+        """Apply a measurement to the state and return numerical results."""
+        # measurements can sometimes create intermediary mps, but those intermediaries will not work with capture enabled
+        disable()
+        try:
+            if self.shots:
+                return self.measurement_class(self.state).measure_with_samples(
+                    [measurement], self.shots
+                )
+            return self.measurement_class(self.state).measurement(measurement)
+        finally:
+            enable()
+
+
+@LightningInterpreter.register_primitive(measure_prim)
+def _(self, *invals, reset, postselect):
+    mp = MidMeasureMP(invals, reset=reset, postselect=postselect)
+    mcms = {}
+    self.state.apply_operations([mp], mid_measurements=mcms)
+    return mcms[mp]
+
+
+# pylint: disable=unused-argument
+@LightningInterpreter.register_primitive(adjoint_transform_prim)
+def _(self, *invals, jaxpr, n_consts, lazy=True):
+    # TODO: requires jaxpr -> list of ops first
+    raise NotImplementedError
+
+
+# pylint: disable=too-many-arguments
+@LightningInterpreter.register_primitive(ctrl_transform_prim)
+def _(self, *invals, n_control, jaxpr, control_values, work_wires, n_consts):
+    # TODO: requires jaxpr -> list of ops first
+    raise NotImplementedError
+
+
+# pylint: disable=too-many-arguments
+@LightningInterpreter.register_primitive(for_loop_prim)
+def _(self, start, stop, step, *invals, jaxpr_body_fn, consts_slice, args_slice):
+    consts = invals[consts_slice]
+    init_state = invals[args_slice]
+
+    res = init_state
+    for i in range(start, stop, step):
+        res = copy(self).eval(jaxpr_body_fn, consts, i, *res)
+
+    return res
+
+
+# pylint: disable=too-many-arguments
+@LightningInterpreter.register_primitive(while_loop_prim)
+def _(self, *invals, jaxpr_body_fn, jaxpr_cond_fn, body_slice, cond_slice, args_slice):
+    consts_body = invals[body_slice]
+    consts_cond = invals[cond_slice]
+    init_state = invals[args_slice]
+
+    fn_res = init_state
+    while copy(self).eval(jaxpr_cond_fn, consts_cond, *fn_res)[0]:
+        fn_res = copy(self).eval(jaxpr_body_fn, consts_body, *fn_res)
+
+    return fn_res
+
+
+@LightningInterpreter.register_primitive(cond_prim)
+def _(self, *invals, jaxpr_branches, consts_slices, args_slice):
+    n_branches = len(jaxpr_branches)
+    conditions = invals[:n_branches]
+    args = invals[args_slice]
+
+    for pred, jaxpr, const_slice in zip(conditions, jaxpr_branches, consts_slices):
+        consts = invals[const_slice]
+        if pred and jaxpr is not None:
+            return copy(self).eval(jaxpr, consts, *args)
+    return ()
```

Comment:
Does every device interpreter need to register these custom primitives? They seem redundant.. Could they be derived from the base interpreter class?

Reply:
So for the default version in `PlxprInterpreter`, I decided to go with default higher order implementations that are "nested", as in we convert the jaxpr using the interpreter, and then rebind the Higher Order Primitive.  This is the kind of behaviour we would want to transformations.

But for device execution, we need these "flattened" versions with for, while, and cond.  I am gathering these definitions into a `qml.capture.FlattenedHigherOrderPrimitives`, so after that PR does in we should be able to do:
```
LightningInterpreter._primitive_registrations.update(qml.capture.FlattenedHigherOrderPrimitives)
```

See [(PR #6343)](https://github.com/PennyLaneAI/pennylane/pull/6343/files)